torch/_refs/__init__.py

```diff
@@ -4647,6 +4648,48 @@ def triu_indices(
     )
 
 
+@register_decomposition(torch.ops.aten.bucketize)
+@out_wrapper()
+def bucketize(
+    a: TensorLikeType,
+    boundaries: TensorLikeType,
+    *,
+    out_int32: bool = False,
+    right: bool = False,
+):
+
+    out_dtype = torch.int32 if out_int32 else torch.int64
+    n_boundaries = boundaries.shape[-1]
+
+    # can't use int32 as indexes, so we have to do all computations with int64 and convert at the end
+    start = torch.zeros(a.shape, device=a.device, dtype=torch.int64)
+    end = start + n_boundaries
+    # Max depth of the binary search
+    niters = int(math.log2(n_boundaries)) + 1
+    if not (right):
+        # equivalent to std::upper_bound
+        for _ in range(niters):
+            cond_update = start < end
+            # start might end up pointing to 1 past the end, we guard against that
+            mid = torch.where(cond_update, start + (end - start) // 2, 0)
```

Comment:
Similarly, on the first iteration of the algorithm is this `where` necessary or is it known how to populate `mid`?

Reply:
Not that the most optimal algorithm needs to be part of this PR, but if there are optimizations we suspect we can make then I suggest we at least add a TODO comment highlighting possible optimizations

Reply:
Sure, see reply above.